index.js

```diff
@@ -1,43 +1,41 @@
-const jwt = require('jsonwebtoken')
-const GitHubApi = require('github')
-
-module.exports = function ({id, cert, debug = false}) {
-  function asApp () {
-    const github = new GitHubApi({debug})
-    github.authenticate({type: 'integration', token: generateJwt(id, cert)})
-    // Return a promise to keep API consistent
-    return Promise.resolve(github)
-  }
+const jsonwebtoken = require('jsonwebtoken')
 
-  // Authenticate as the given installation
-  function asInstallation (installationId) {
-    return createToken(installationId).then(res => {
-      const github = new GitHubApi({debug})
-      github.authenticate({type: 'token', token: res.data.token})
-      return github
-    })
-  }
+function jwt (id, pem) {
+  console.log({id, pem})
 
-  // https://developer.github.com/early-access/integrations/authentication/#as-an-installation
-  function createToken (installationId) {
-    return asApp().then(github => {
-      return github.apps.createInstallationToken({
-        installation_id: installationId
-      })
-    })
+  const payload = {
+    iat: Math.floor(new Date() / 1000),       // Issued at time
+    exp: Math.floor(new Date() / 1000) + 60,  // JWT expiration time
+    iss: id                                   // Integration's GitHub id
   }
 
-  // Internal - no need to exose this right now
-  function generateJwt (id, cert) {
-    const payload = {
-      iat: Math.floor(new Date() / 1000),       // Issued at time
-      exp: Math.floor(new Date() / 1000) + 60,  // JWT expiration time
-      iss: id                                   // Integration's GitHub id
-    }
+  // Sign with RSA SHA256
+  return jsonwebtoken.sign(payload, pem, {algorithm: 'RS256'})
+}
 
-    // Sign with RSA SHA256
-    return jwt.sign(payload, cert, {algorithm: 'RS256'})
-  }
+module.exports = ({id, pem}) => {
+  return octokit => {
+    octokit.jwt = jwt.bind(null, id, pem)
+
+    // Save the original octokit authenticate method
+    const authenticate = octokit.authenticate
 
-  return {asApp, asInstallation, createToken}
+    // Authenticate this instance as the GitHub App
+    // FIXME: this will only work for requests within the next 60 sec
+    authenticate({ type: 'app', token: octokit.jwt() })
+
+    octokit.authenticate = async (options) => {
+      if (options.type === 'installation') {
+        octokit.authenticate({ type: 'app', token: octokit.jwt() })
+
+        const { data: { token } } = await octokit.apps.createInstallationToken({
+          installation_id: options.id
+        })
+
+        return authenticate({ type: 'token', token })
+      } else {
+        return authenticate(options)
+      }
+    }
```

Comment:
instead of overwriting the `octokit.authenticate` method, you should be able to use `octokit.hook.before('request', setAuthHeader)` to set the authentication headers correctly. The hooks can be async

Reply:
Thoughts then on what this API should be? What should I call if I want to authenticate as a specific installation?

```js
octokit.authenticate({ type: 'installation', id: 12345})
```

If I skip overloading the authenticate method, I suppose I could just use the registering of the plugin as the signal:

```js
const app = require('@octokit/github-app')({
  id: 1234,
  pem: fs.readFileSync('private-key.pem', 'UTF-8')
})

// authenticate as the app
octokit.plugin(app)

// authenticate as an installation
octokit.plugin(app.installation(1234))
```

Reply:
Let me get back to that once I ship the next `@octokit/rest`, ok?

I think authenticating as app/installation should be built into octokit itself.

But in terms of the plugin API, I’d say it should end up looking something like this

```js
// my-plugin.js
return (client) => {
  client.asInstallation = ({appId, privateKey, installationId}) => {
    const appClient = octokit.authenticate({ type: 'app', token: getJsonWebToken({appId, privateKey})})
    installationClient.hook(async (requestOptions, requestMethod) => {
      // TODO: cache installation, etc
      const installationToken = appClient.apps.createInstallationToken({installation_id: installationId})
      requestOptions.headers.authorization = `token ${installationToken}`
      return requestMethod(requestOptions)
    })
  }
})

// my-app.js
const Octokit = require('octokit').plugin(require('my-plugin'))
const client = new Octokit()
const installation = client.asInstallation({appId, privateKey, installationId})
```

See also https://github.com/octokit/app.js/pull/1 and https://github.com/octokit/octokit.js/blob/initial-version/README.md